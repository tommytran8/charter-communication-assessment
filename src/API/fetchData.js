import axios from "axios";

const fetchData = () => {
  return axios.get(`${window.location.origin}/data`);
};

export default fetchData;
