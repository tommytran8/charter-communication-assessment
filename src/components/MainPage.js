import React, { useEffect, useRef, useState } from "react";
import fetchData from "../API/fetchData";

const getMonthOfTransaction = (transaction) => {
  return new Date(transaction.dateOfTransaction).toLocaleString("default", {
    month: "short",
  });
};

const generateMonthArr = (customerTransactions) => {
  const monthSet = new Set();
  customerTransactions.forEach((transaction) => {
    const monthOfTransaction = getMonthOfTransaction(transaction);
    monthSet.add(monthOfTransaction);
  });
  return monthSet;
};

const calculateTotalRewards = (transaction) => {
  let totalRewardPoints = 0;
  if (transaction.customerExpense > 100) {
    const dollarsSpentOver100 = transaction.customerExpense - 100;
    const dollarsSpentOver50 =
      transaction.customerExpense - dollarsSpentOver100 - 50;
    totalRewardPoints += 2 * dollarsSpentOver100 + 1 * dollarsSpentOver50;
  } else if (transaction.customerExpense > 50) {
    const dollarsSpentOver50 = transaction.customerExpense - 50;
    totalRewardPoints += 1 * dollarsSpentOver50;
  }
  return totalRewardPoints;
};

const MainPage = () => {
  const [customerTransactions, setCustomerTransactions] = useState(null);
  const [customerRewardPoints, setCustomerRewardPoints] = useState({});
  const monthSet = useRef([]);

  useEffect(() => {
    fetchData().then((res) => {
      monthSet.current = Array.from(generateMonthArr(res.data.transactions));
      setCustomerTransactions(res.data.transactions);
      const [month1, month2, month3] = monthSet.current;
      res.data.transactions.forEach((transaction) => {
        setCustomerRewardPoints((prev) => {
          return {
            ...prev,
            [transaction.customerId]: {
              ...prev[transaction.customerId],
              totalPoints: 0,
              pointsForEachMonth: {
                [month1]: 0,
                [month2]: 0,
                [month3]: 0,
              },
            },
          };
        });
      });
    });
  }, []);

  useEffect(() => {
    customerTransactions?.forEach((transaction) => {
      const totalRewardPoints = calculateTotalRewards(transaction);
      const monthOfTransaction = getMonthOfTransaction(transaction);
      setCustomerRewardPoints((prev) => {
        return {
          ...prev,
          [transaction.customerId]: {
            ...prev[transaction.customerId],
            totalPoints:
              prev[transaction.customerId].totalPoints + totalRewardPoints,
            pointsForEachMonth: {
              ...prev[transaction.customerId].pointsForEachMonth,
              [monthOfTransaction]:
                prev[transaction.customerId].pointsForEachMonth[
                  monthOfTransaction
                ] + totalRewardPoints,
            },
          },
        };
      });
    });
  }, [customerTransactions]);

  return (
    <div className="solution-container">
      <table>
        <tbody>
          <tr>
            <th>customerId</th>
            <th>totalPoints</th>
            {monthSet.current ? (
              monthSet.current.map((month) => <th key={month}>{month}</th>)
            ) : (
              <></>
            )}
          </tr>
          {customerRewardPoints ? (
            Object.keys(customerRewardPoints).map((customerId) => {
              return (
                <tr key={customerId}>
                  <td>{customerId}</td>
                  <td>{customerRewardPoints[customerId].totalPoints}</td>
                  {monthSet.current ? (
                    monthSet.current.map((month) => (
                      <td key={month}>
                        {
                          customerRewardPoints[customerId].pointsForEachMonth[
                            month
                          ]
                        }
                      </td>
                    ))
                  ) : (
                    <></>
                  )}
                </tr>
              );
            })
          ) : (
            <></>
          )}
        </tbody>
      </table>
    </div>
  );
};

export default MainPage;
