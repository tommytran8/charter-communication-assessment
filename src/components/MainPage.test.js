import { render, screen } from "@testing-library/react";
import MainPage from "./MainPage";
import axios from "axios";

import fetchData from "../API/fetchData";

jest.mock("axios");

describe("MainPage", () => {
  describe("when API call is successful", () => {
    test("should return dataset of transactions", async () => {
      const transactions = [
        {
          customerId: 1,
          customerName: "Andrew Jackson",
          customerExpense: 105,
          dateOfTransaction: "11/10/2023",
        },
        {
          customerId: 2,
          customerName: "Jack Daniels",
          customerExpense: 200,
          dateOfTransaction: "12/03/2022",
        },
        {
          customerId: 3,
          customerName: "John Paul",
          customerExpense: 59,
          dateOfTransaction: "12/20/2022",
        },
      ];
      axios.get.mockResolvedValueOnce(transactions);

      const result = await fetchData();

      expect(axios.get).toHaveBeenCalledWith(`${window.location.origin}/data`);
      expect(result).toEqual(transactions);
    });
    test("render page with dataset", async () => {
      const transactions = [
        {
          customerId: 1,
          customerName: "Andrew Jackson",
          customerExpense: 105,
          dateOfTransaction: "11/10/2023",
        },
        {
          customerId: 2,
          customerName: "Jack Daniels",
          customerExpense: 200,
          dateOfTransaction: "12/03/2022",
        },
        {
          customerId: 3,
          customerName: "John Paul",
          customerExpense: 59,
          dateOfTransaction: "12/20/2022",
        },
      ];
      axios.get.mockResolvedValueOnce(transactions);
      const result = await fetchData();
      expect(axios.get).toHaveBeenCalledWith(`${window.location.origin}/data`);
      expect(result).toEqual(transactions);
      render(<MainPage />);
      expect(screen.getByText("/Dec/i")).toBeInTheDocument();
    });
  });
});
