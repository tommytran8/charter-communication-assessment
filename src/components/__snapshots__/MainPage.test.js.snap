// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="solution-container"
  >
    <table>
      <tbody>
        <tr>
          <th>
            customerId
          </th>
          <th>
            totalPoints
          </th>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
